src/adapters/notifications/NotificationCenterAdapter.ts

```diff
@@ -819,4 +692,259 @@ export class NotificationCenterAdapter implements NotificationAdapterInterface {
       },
     });
   }
+
+  async broadcastNotification(
+    params: BroadCastNotificationInputParams,
+  ): Promise<void> {
+    const {
+      broadCastTitle,
+      link,
+      linkTitle,
+      text,
+      sendEmail,
+      broadCastNotificationId,
+    } = params;
+    let allUserFetched = false;
+    const take = 100;
+    let skip = 0;
+    const trackIdPrefix = `broadCast-${broadCastNotificationId}`;
+    while (!allUserFetched) {
+      const { users } = await findAllUsers({ take, skip });
+      if (users.length === 0) {
+        allUserFetched = true;
+        break;
+      }
+      skip += users.length;
+      const queueData: SendBatchNotificationBody = [];
+      for (const user of users) {
+        queueData.push({
```

Comment:
I don't think this is necessary here @mohammadranjbarz. We can offload all the work to the notification center, because this would spam the notification center with many requests, it might hit a rate limit (not sure if we have with nginx).

Remember the notification center also have all users registered with their wallet address. 

Simply send a request with the type: "Broadcast" and the data, and handle the segment loop in the notification center in multiple queues. (I think we would need to implement the queue system in notification center that we don't currently have).

For the notifications to insert in the DB, I think we can insert all the values in one single query in the notification center.
https://stackoverflow.com/questions/24344783/insert-into-for-all-users-who-are-in-another-table

So to summarize:
1. Send 1 request with the type of notification **broadcast** and the data. (this enqueues the segment events enqueuer looper).
2. Run the query to insert all values to all users in a single query.  (I can help with this)
3. The segment events enqueuer, loops and enqueues each individual user notification to be sent with segment (in its own queue).

Its a lot more work, but its more efficient.

Reply:
BUT!!!!!!!!!! ***************************

If we are short in time, we can use this to start, but the loop do it inside a Queue not in the main thread. Because We had issues before where doing heavy lifting in an endpoint (like not paginating) and the backend slows down.

(this is my limited knowledge in js, for instance since it's single threaded does bull spawn new threads? or uses the same one?) (if its the same thread then doing the loop in a queue won't do anything performance wise, and you can ignore me)

Reply:
Thanks @CarlosQ96  for your comments, I remember you had mentioned on the specs of issue as well

During the holidays I and Amin talked about that and we came to this conclusion:
1. We don't have all of users in notification-center, we just have the user's info that we already sent them at least 1 notification
2. We don't have emails of users there
3. We implemented https://staging.notification.giveth.io/docs/#/Notification/SendBulkNotification s we can send 100 notification in one request, so if all of our users are 2,000 we can sen notification for them with only 20 requests  and also we queue the requests so it would not make lots of load on notification-center
4. probably we don't use this feature too much, maybe just few times in a month (maybe less) so I think a minimal implementation can be enough for that

So I think yeah current implementation might be good at the beginning, do you agree with me?